### PR TITLE
Add schedule classes

### PR DESCRIPTION
**Pull request by jneeven**
These are useful for combining several schedules, e.g. linear warmup and cosine decay, in a clean way that also ensures the hyperparameter is logged to tensorboard. The `Schedule` class accepts any callable, including keras schedules like `keras.optimizers.schedules.PolynomialDecay`, or a custom function. All of these have been tested on object detection and are very convenient there.

Using these with the `HyperparameterScheduler` is simple:
```python
     if self.learning_rate_schedule:
            callbacks.append(
                lq.callbacks.HyperparameterScheduler(
                    self.learning_rate_schedule,
                    "lr",
                    update_freq=self.learning_rate_schedule.update_freq,
                )
            )
```

Example for linear warmup + cosine decay:
```python
    @cached_property
    def main_schedule(self):
        main_epochs = self.epochs - self.warmup_duration
        # The -1 is necessary to actually reach 0.
        return utils.Schedule(
            schedule=keras.experimental.CosineDecay(
                self.initial_lr,
                (self.steps_per_epoch * main_epochs) - 1,
                name="main_schedule",
            )
        )

    @cached_property
    def learning_rate_schedule(self):
        warmup_schedule = utils.Schedule(
            schedule=keras.optimizers.schedules.PolynomialDecay(
                0.0, self.steps_per_epoch * self.warmup_duration, self.initial_lr,
            )
        )
        return utils.CombinedSchedule(
            schedules=[warmup_schedule, self.main_schedule],
            starting_times=[self.warmup_duration * self.steps_per_epoch],
        )
```

It's a draft because it needs more tests etc, but feel free to review the current code!

TODO:
- [ ] Write unit tests
- [ ] Add docstrings
- [ ] Add some code examples in docstrings

### CODE REVIEW

**Comment by jneeven** [comment]
@lgeiger  any ideas on this? Do we want this in Larq or elsewhere?

**Comment by lgeiger** [comment]
> @lgeiger any ideas on this? Do we want this in Larq or elsewhere?

I personally would prefer having it elsewhere, since it is not directly related to quantization. To keep the scope of larq focused.

**Comment by jneeven** [comment]
> > @lgeiger any ideas on this? Do we want this in Larq or elsewhere?
> 
> I personally would prefer having it elsewhere, since it is not directly related to quantization. To keep the scope of larq focused.

Sounds good, shall I move it to research temp?
